edgedb/codegen/sertypes.py

```diff
@@ -0,0 +1,482 @@
+    tid = uuid.UUID(bytes=desc.read_bytes(16))
+
+    if t == CTYPE_SET:
+        pos = desc.read_ui16()
+        return SetDesc(tid=tid, subtype=codecs_list[pos])
+
+    elif t == CTYPE_SHAPE or t == CTYPE_INPUT_SHAPE:
+        els = desc.read_ui16()
+        fields = {}
+        flags = {}
+        cardinalities = {}
+        fields_list = []
+        for idx in range(els):
+            flag = desc.read_ui32()
+            cardinality = Cardinality(desc.read_bytes(1)[0])
+            name = desc.read_len32_prefixed_bytes().decode()
+            pos = desc.read_ui16()
+            codec = codecs_list[pos]
+            if t == CTYPE_INPUT_SHAPE:
+                fields_list.append((name, codec))
+                fields[name] = idx, codec
+            else:
+                fields[name] = codec
+            flags[name] = flag
+            if cardinality:
+                cardinalities[name] = cardinality
+        args = dict(
+            tid=tid,
+            flags=flags,
+            fields=fields,
+            cardinalities=cardinalities,
+        )
+        if t == CTYPE_SHAPE:
+            return ShapeDesc(**args)
+        else:
+            return InputShapeDesc(fields_list=fields_list, **args)
+
+    elif t == CTYPE_BASE_SCALAR:
+        return BaseScalarDesc(tid=tid)
+
+    elif t == CTYPE_SCALAR:
+        pos = desc.read_ui16()
+        return ScalarDesc(tid=tid, subtype=codecs_list[pos])
+
+    elif t == CTYPE_TUPLE:
+        els = desc.read_ui16()
+        fields = []
+        for _ in range(els):
+            pos = desc.read_ui16()
+            fields.append(codecs_list[pos])
+        return TupleDesc(tid=tid, fields=fields)
+
+    elif t == CTYPE_NAMEDTUPLE:
+        els = desc.read_ui16()
+        fields = {}
+        for _ in range(els):
+            name = desc.read_len32_prefixed_bytes().decode()
+            pos = desc.read_ui16()
+            fields[name] = codecs_list[pos]
+        return NamedTupleDesc(tid=tid, fields=fields)
+
+    elif t == CTYPE_ENUM:
+        els = desc.read_ui16()
+        names = []
+        for _ in range(els):
+            name = desc.read_len32_prefixed_bytes().decode()
+            names.append(name)
+        return EnumDesc(tid=tid, names=names)
+
+    elif t == CTYPE_ARRAY:
+        pos = desc.read_ui16()
+        els = desc.read_ui16()
+        if els != 1:
+            raise NotImplementedError(
+                "cannot handle arrays with more than one dimension"
+            )
+        dim_len = desc.read_i32()
+        return ArrayDesc(tid=tid, dim_len=dim_len, subtype=codecs_list[pos])
+
+    elif t == CTYPE_RANGE:
+        pos = desc.read_ui16()
+        return RangeDesc(tid=tid, inner=codecs_list[pos])
+
+    elif t[0] >= 0x80 and t[0] <= 0xFF:
+        # Ignore all type annotations.
+        desc.read_len32_prefixed_bytes()
+        return None
+
+    else:
+        raise NotImplementedError(
+            f"no codec implementation for EdgeDB data class {t}"
+        )
```

Comment:
This looks sufficiently similar to how the codecs tree is built and that code is already in this package. Can we potentially reuse it instead of copy/pasting this?

Reply:
You can traverse the tree of codecs and use singledispatch to implement custom logic needed here. IDK, I'd do that if that's not too hard; otherwise this is fine too (albeit this copy/paste still looks worrisome).

Reply:
This copy/paste was intended to be deleted after we can do the same in the server-side `sertypes.py` as a language server in the (far) future. This is also why the `raw_parse()` is added below so that we can "recover" the `sertypes.py` codecs.

Reply:
> This is also why the raw_parse() is added below so that we can "recover" the sertypes.py codecs.

I don't understand what do you mean by "recover" here...

Reply:
Oh sorry, I meant in the future language server, we won't need to serialize and deserialize the type descriptors, just 1) compile/analyze the query and get `sertypes.py` codecs objects in memory, 2) generate the JSON schema directly using the same code as we are having in this PR here in `sertypes.py`. However we don't have the language server now, so I'm simply copying over part of the `sertypes.py` that re-loads ("recover") the codec objects from serialized type descriptors, and do the (2) step as what we will do in the future language server. An up-side is that we don't have to interfere/hack too much into the C code in edgedb-python now (so that it'll be easier to migrate to the language server in the future), while the downside is to maintain two copies of `sertypes.py` until the language server is developed (but I guess we don't usually modify the codes, so...).